### DISCUSSION TITLE
Repeated calls to cache.updateQuery do not perform cumulative effects

### DISCUSSION THREAD

**Discussion by codazzo** [urql-graphql/urql#1073] [Q&A] [answered]
Hi! As I mentioned in https://github.com/FormidableLabs/urql/discussions/1068#discussioncomment-101293 I'm trying to use urql for a scenario where I expect users of a native app to be offline for long enough stretches of time wherein they can perform a number of mutations, each of which would create a "Recommendation" entity.

For this I'm using a custom updater

```ts
offlineExchange({
    storage, // <--- I created this using AsyncStorage
    schema,
    updates: {
        Mutation: {
            createRecommendation: (result, args, cache, info) => {
                const recommendation = .... // builds recommendation from args
                const fieldId = ... // gets field Id from args

                cache.updateQuery({ query: QUERY_RECOMMENDATIONS, variables: { fieldId } }, (data) => {
                    data.recommendations.push(recommendation);
                    return data;
                });
            },
        },
    },
    optimistic: {
        createRecommendation: ... // creates a Recommendation entry
    }
});
```

This works the first time: even if the device is disconnected, I can see the newly-created recommendation if I use the client to query `QUERY_RECOMMENDATIONS`. Upon following attempts though, keeping the device offline, `data.recommendations` in the updater function passed to `cache.updateQuery` does not include any recommendation that was previously created _while the device was offline_. As a consequence, if I query `QUERY_RECOMMENDATIONS` after the mutation, I always only see only the latest recommendation that was created offline (plus all of the recommendations that were created while the device was connected).

Is there a way these offline updates can be cumulative, i.e. I could create one recommendation after the other and see them all while the device is offline?

Thanks! 🙏

**Comment by kitten** [comment, marked as answer]
That's pretty odd! I'm trying to think of what could be causing this, because optimistic updates should always allow you to cumulatively add to a list already. So let's debug this and maybe I can write a new test for our test suite to make sure that this is indeed broken and then subsequently fix it if it's indeed buggy.

So on the second optimistic `createRecommendation` call of `updates.Mutation.createRecommendation`, have you verified that it's indeed called, that the list is missing the last recommendation, or whether any other weird behaviour already occurs there?

You said that it's called but lacking the previous recommendation; that's all while the device was offline and only optimistic updates were made, right?

**Reply by kitten** [reply]
Ok, I think I actually already have a nice and simple reproduction in our test suite done, so I'll look into this

**Reply by codazzo** [reply]
>  So let's debug this and maybe I can write a new test for our test suite to make sure that this is indeed broken and then subsequently fix it if it's indeed buggy.

That's fantastic and much appreciated!

> have you verified that it's indeed called

Yup! Both the optimistic mutation and the custom updater are called (I verified this via devtools).

> the list is missing the last recommendation

It is. Curiously enough, if I stop at a breakpoint on the line before `cache.updateQuery`, then `cache.readQuery` will also not include any entities created while offline but if I run `client.query(QUERY_RECOMMENDATIONS, { fieldId }).toPromise().then(console.log)` on the devtools console I see all entries.

> You said that it's called but lacking the previous recommendation; that's all while the device was offline and only optimistic updates were made, right?

I can reproduce this behaviour in both of these scenarios:

* iOS simulator pointing to API running on localhost, with API process killed
* real device with no connectivity

**Reply by kitten** [reply]
Alright, I've got a fix ready for this. As it turned out _we did this on purpose_ and I'm not sure anymore why we deemed cumulative optimistic updates unsafe, because they shouldn't be. So if you'd like to test these changes I'd be super grateful 🙏  https://github.com/FormidableLabs/urql/pull/1074

You can test out this set of changes by using the package that CodeSandbox CI builds for us. It'll give you a link to a temporary package once it's done.

**Reply by codazzo** [reply]
That's amazing 😍  I will test this right away. I've been running `urql` via `yarn link` for a while anyway to poke around the source so it should just be a `fetch` and `checkout` away. Will report!

**Reply by kitten** [reply]
Ok, note that this is a Graphcache issue, so it's not related to `urql`. Also btw, there may be some issues with duplicate dependencies when you `yarn link` so make sure you're not seeing weird unrelated issues 😅 Anyway, to test this out you can also install Graphcache in your `package.json` from `https://pkg.csb.dev/FormidableLabs/urql/commit/16c0f581/@urql/exchange-graphcache`. So if you set the `@urql/exchange-graphcache` version to that, you'll get this fix.

**Reply by codazzo** [reply]
I can confirm that the updates were applied cumulatively and the mutations were performed when the device got back online. Thanks for the stellar support!

**Reply by kitten** [reply]
This has also just been released! Can't wait to hear more about how you get on! 🙌  https://github.com/FormidableLabs/urql/releases/tag/%40urql%2Fexchange-auth%400.1.1
